components/search/BookingInfoSearch/BookingInfoSearchUIComplete.tsx

```diff
@@ -1,57 +1,97 @@
+const BookingInfoUISearchComplete: React.FC<
+  BookingInfoUISearchCompleteProps
+> = ({ bookingInfo, setBookingInfo }) => {
+  const [tempBookingInfo, setTempBookingInfo] = useState(bookingInfo);
+  const [isDateValid, setIsDateValid] = useState(true);
+  const [isValid, setIsValid] = useState(true);
+
   const handleDateChange = (dates: {
     checkinDate: string;
     checkoutDate: string;
     numDays: number;
   }) => {
-    setBookingInfo({ ...bookingInfo, ...dates });
+    setTempBookingInfo((prev) => ({ ...prev, ...dates }));
+  };
+
+  const handleAdultsNumberChange = (adults: number | null) => {
+    if (adults !== null) {
+      setIsValid(true);
+      setTempBookingInfo((prev) => ({ ...prev, adultsNumber: adults }));
+    } else {
+      setIsValid(false);
+    }
   };
 
-  const handleAdultsNumberChange = (adults: number) => {
-    setBookingInfo({ ...bookingInfo, adultsNumber: adults });
+  const handleApplyFilters = () => {
+    if (isValid) {
+      setBookingInfo(tempBookingInfo);
+    }
+  };
+
+  const handleRevertChanges = () => {
+    const confirmRevert = window.confirm(
+      "Are you sure you want to revert your booking changes?"
+    );
+    if (confirmRevert) {
+      setTempBookingInfo(bookingInfo);
+      setBookingInfo(tempBookingInfo);
+    }
   };
 
   return (
-    <div >
+    <div>
       <h2 className="text-2xl font-bold">Booking Information</h2>
-
       <div className="gap-5 join">
-        {/* Booking Date Range Input */}
-          <div className="join-item">
-            <h3 className="text-primary font-semibold">Booking Dates</h3>
-            <DatePickerWithRange
-              onChange={handleDateChange} // Update parent state when date range changes
-              defaultNumDays={bookingInfo.numDays} // Default number of days
-            />
-          </div>
-          {/* Adults Number Input */}
-          <div className="join-item">
-            <AdultsNumberInput
-              selectedNumber={bookingInfo.adultsNumber}
-              onChange={handleAdultsNumberChange}
-            />
-          </div>
+        <div className="join-item">
+          <h3 className="text-primary font-semibold">Booking Dates</h3>
+          <DatePickerWithRange
+            bookingInfo={bookingInfo}
+            onChange={handleDateChange}
+            onValidationChange={setIsDateValid}
+          />
+        </div>
+        <div className="join-item">
+          <AdultsNumberInput
+            selectedNumber={tempBookingInfo.adultsNumber}
+            onChange={handleAdultsNumberChange}
+          />
+          {!isValid && (
+            <p className="text-red-500 text-sm">
+              Adults number is invalid. Please enter a valid number.
+            </p>
+          )}
+        </div>
       </div>
-
+      {(bookingInfo.checkinDate !== tempBookingInfo.checkinDate ||
+        bookingInfo.checkoutDate !== tempBookingInfo.checkoutDate ||
+        bookingInfo.adultsNumber !== tempBookingInfo.adultsNumber) && (
```

Comment:
This could be simplified by checking if `bookingInfo !== tempBookingInfo` just be careful to check for deep equality instead of references.
An example is `JSON.stringify(bookingInfo, Object.keys(bookingInfo).sort()) == JSON.stringify(tempBookingInfo, Object.keys(tempBookingInfo).sort())`

Reply:
Will probably keep this. bookinginfo is of type searchParams (which includes a ton of other fields). I'm only looking for those specific fields like `checkInDate`, `checkoutDate`, and `adultsNumber`.